### ISSUE TITLE
When the indexer is restarted, it re-saves the blocks that have already been stored.

### ISSUE THREAD

**Issue by hhhroot** [blockscout/blockscout#9753]
### Description

When the indexer is restarted, it re-saves blocks that have already been stored. It does not save all the blocks again but attempts to save the same blocks multiple times. (For example, when turned off and on 5 times, it retries blocks 123, 1234, 5678, 9999 each time.)
I wonder why this is happening. I've carefully examined various tables like blocks, transactions, token_transfers, address_token_balances, but I couldn't find any incorrect data. Even after reindexing is complete, if I turn off the indexer and then turn it back on, it reindexes again. (It seems to reproduce if I turn it off and wait for about 5 minutes before turning it back on.)

example query result - all data except updated_at is the same.
```SQL
select inserted_at, updated_at
from transactions
where block_number = 925626;
```

inserted_at | updated_at
2024-02-01 02:39:50.042496 | 2024-03-29 09:06:59.947358
2024-02-01 02:39:50.042496 | 2024-03-29 09:06:59.947358
2024-02-01 02:39:50.042496 | 2024-03-29 09:06:59.947358
2024-02-01 02:39:50.042496 | 2024-03-29 09:06:59.947358

inserted_at | updated_at
2024-02-01 02:39:50.042496 | 2024-03-29 10:15:21.936
2024-02-01 02:39:50.042496 | 2024-03-29 10:15:21.936
2024-02-01 02:39:50.042496 | 2024-03-29 10:15:21.936
2024-02-01 02:39:50.042496 | 2024-03-29 10:15:21.936

inserted_at | updated_at
2024-02-01 02:39:50.042496 | 2024-03-29 11:35:55.382923
2024-02-01 02:39:50.042496 | 2024-03-29 11:35:55.382923
2024-02-01 02:39:50.042496 | 2024-03-29 11:35:55.382923
2024-02-01 02:39:50.042496 | 2024-03-29 11:35:55.382923

**What does the indexer see and try to store the data back?**

### Type of the installation

Docker-compose

### Type of the JSON RPC archive node

geth

### Type of the chain

optimism

### Backend version

v6.1.0

### Elixir & Erlang/OTP versions

elixir:1.14.5-erlang-24.2.2-alpine-3.18.2

**Comment by hhhroot** [comment]
I investigated the issue in more detail. You can get the blocks that are trying to be reindexed using the query below.
```SQL
SELECT DISTINCT block_number
FROM logs
WHERE 
  first_topic = '\x4a39dc06d4c0dbc64b70af90fd698a233a518aa5d07e595d983b8c0526c8f7fb' 
  AND data = '\x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000';
```
It seems that the indexer considers blocks with this log to be incorrect and attempts to reindex them. I believe that this value should be handled as an exception in the indexer.

**Comment by sl1depengwyn** [comment]
These logs seems to be related to ERC-1155 batch transfer, do you see any related errors in logs?
Try to look for this pattern `Unknown token transfer format`

**Comment by hhhroot** [comment]
There are logs that you mentioned, but it's not the case I'm talking about. (The topic and data are different) I think the indexer recognizes this log as a correct token transfer, but it doesn't actually send anything, so it seems to be not handled. Can you look at [this](https://etherscan.io/tx/0xc61c2c6ab5984a129eb038d9d8f4a1b062082af39cd671d7bc07c328812a452a#eventlog)? Looking at the data of the log, you can see that ids and values are empty.

**Comment by hhhroot** [comment]
@sl1depengwyn @vbaranov

**Comment by sl1depengwyn** [comment]
Actually, we ignore such empty transfers and store nothing for them, but this does not trigger indexer to reindex blocks, I checked this locally and on eth mainnet instance, this block wasn't updated there for a long period
The reason is not in this particular log I think, could you please try to look for any other similarities between those blocks that are imported on restart on your end?

**Comment by hhhroot** [comment]
I've looked over various data over the course of several days, but I haven't found any differences.

My case is about encountering a problem when restarting the indexer. (I stopped and then restarted the Docker Compose for about 5 minutes.) Did you actually stop and then restart the indexer? If you don't turn the indexer back on, as you said, it doesn't reindex separately. - I think there would be a problem if you take a snapshot of the database and start from there again. Can you try it?

**Comment by sl1depengwyn** [comment]
Yeah, I've restarted the application and eth mainnet instance restarts periodically, but such logs do not trigger indexer to re-import the blocks

**Comment by sl1depengwyn** [comment]
I think it would be helpful if you can share snapshot of your database if it's possible?

**Comment by hhhroot** [comment]
How do I share snapshots? And can you check that you are trying to reindex with this query? Updates occur on specific tables, not all tables.
```SQL
WITH weird_blocks AS (
  SELECT DISTINCT block_number
  FROM logs
  WHERE 
    first_topic = '\x4a39dc06d4c0dbc64b70af90fd698a233a518aa5d07e595d983b8c0526c8f7fb' 
    AND data = '\x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'
  ORDER BY block_number
  LIMIT 10
)
SELECT atb.block_number, value_fetched_at, inserted_at, updated_at
FROM address_token_balances atb JOIN weird_blocks wb ON atb.block_number = wb.block_number;
```

**Comment by hhhroot** [comment]
@sl1depengwyn

**Comment by hhhroot** [comment]
@sl1depengwyn bump
can you try the sql?

**Comment by vbaranov** [comment]
Please retry with the latest version of the backend.
